src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/BufferSegment.netcoreapp.cs

```diff
@@ -0,0 +1,43 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace System.IO.Pipelines
+{
+    internal sealed partial class BufferSegment : ReadOnlySequenceSegment<byte>
+    {
+        public void ResetMemory()
+        {
+            object? memoryOwner = Interlocked.Exchange(ref _memoryOwner, null);
+            if (memoryOwner is null)
+            {
+                // Already returned block; this can happen from benign race between Advance and Complete
+                // as block return is done outside a lock.
+                return;
+            }
+
+            // Use a fast exact type check rather than checking the inheritance hierarchy
+            // or following the interface mapping.
+            if (memoryOwner.GetType() == typeof(byte[]))
```

Comment:
We could call GetType() in the netstandard version too, right? Does the normal cast vs Unsafe.As make it less worthwhile?

Reply:
Normal cast means you use do a `IsInstanceOfClass` check again for the ArrayPool and `IsInstanceOfInterface` for the MemoryPool, so it would just add extra overhead.

The PR in this area originally used `Unsafe`, however couldn't workout how to get the packaging to work so backed it out https://github.com/dotnet/corefx/pull/39643

Reply:
This change halves the number of `IsInstanceOfInterface` calls for Json Platform

Before
![image](https://user-images.githubusercontent.com/1142958/82518498-6ccabb00-9b17-11ea-9ce1-57ac743fc302.png)

After
![image](https://user-images.githubusercontent.com/1142958/82518539-83711200-9b17-11ea-8afd-b43122425b94.png)